Improve check errors

erikmaarten opened the pull request:
Fixes #4882.

In addition it improves a few of the other error messages.

The changes to `Check.numeric.minimum` and `Check.numeric.maximum` breaks the existing usage of these, so this needs to be fixed before merging. 

In hindsight I think it would have been better to make the name the first parameter of all of the `Check` methods, but to keep parameter order consistent with what we already have in all `typeOf` methods I put the `name` parameter after the variable being tested.

lilleyse commented on Source/Core/Check.js:
```diff
@@ -49,31 +49,35 @@ define([
-        Check.typeOf.number(maximum);
+    Check.numeric.maximum = function (test, name, maximum) {
+        Check.typeOf.number(test, 'test');
+        Check.typeOf.string(name, 'name');
```
Is this check needed? It isn't done in the other functions.

erikmaarten replied:
It's more helpful here than in the `typeOf` checks since there is an additional parameter to potentially misplace here, and the old interface had different parameters. I guess it's probably most line with the fail-early strategy to type check all parameters, as long as there's a chance to catch a problem early and as long as we can afford it. What do you think?

lilleyse replied:
I guess one downside is that the error messages for these checks usually indicate that invalid arguments were sent to cesium functions. However `name` is not an argument sent in by the user, so if `name` is incorrect it's a cesium library coding error as opposed to a user coding error.

erikmaarten replied:
Good point. And actually, throwing an error referring to a variable called `'test'` doesn't make much sense to the end user either, it would be better to use the name of the variable being tested. So `Check.typeOf.number(test, 'test')` would be `Check.typeOf.number(test, name)`.

erikmaarten replied:
And looking at the how `Check.numeric` `minimum` and `maximum` are currently used, it doesn't look like they should actually check that the parameters are of the right types because this is currently always done before calling `Check.numeric.minimum` or `Check.numeric.maximum` anyway.

mramato replied:
It would probably be better to keep the check in minimum/maximum and not check type first in the calling code.  This way it's impossible to not call minimum/maximum with non-numeric data by accident.

erikmaarten commented:
@lilleyse I've updated this and removed the parameter type checks in `Check.numeric.minimum` and `Check.numeric.maximum`. 

It would be nice to be able combine typeof number and range checks, like `Check.typeOf.number(value, 'value').maximum(4)` or even `Check.typeOf.number(value, 'value').range(4, 6)`. That way we could avoid a fair bit of repetition, like this
```js
Check.typeOf.number(row, 'row');
Check.numeric.minimum(row, 0);
Check.numeric.maximum(row, 3);

Check.typeOf.number(column, 'column');
Check.numeric.minimum(column, 0);
Check.numeric.maximum(column, 3);
```
would become

```js
Check.typeOf.number(row, 'row').range(0, 3);
Check.typeOf.number(column, 'column').range(0, 3);
```

We removed the initial range check function because we didn't know if it would actually be that useful, but it looks like there are a fair number of places where this kind of pattern is used.

shehzan10 commented:
> Check.typeOf.number(row, 'row').range(0, 3);
> Check.typeOf.number(column, 'column').range(0, 3);

I think one of the problems with this, as well as the previous `minimum`/`maximum` version, is that there is no obvious information about inclusion/exclusion of the range. If we tend to use one of them always (as we are using inclusion now), this leaves out the option when you do want to exclude one or both of the limits from the range.

I think there were a few lines of code where this made us not use `Check` and use regular if conditions.

What are your thoughts on this?

erikmaarten commented:
@shehzan10 I've considered that too, and we may be able to solve that with clearer naming. (I actually thought `minimum`/`maximum` were semantically clear though, as including the given limit.) 

A range check function could be called `rangeInclusive` or `rangeExclusive`, or if we want it to be more flexible than that (e.g., include one but not the other endpoint), the `range` function could take an object that specifies something like `{upperLimitInclusive: 4, lowerLimitExclusive: 2}` which would translate to the range `(2, 4]`.

Or we could have something like:
```js
Check.typeOf.number(row, 'row').lessThan(3);
Check.typeOf.number(row, 'row').lessThanOrEqual(3);
Check.typeOf.number(row, 'row').lessThan(3).greaterThan(2);
```
where we with skip the range altogether in favor of chaining more specific minimum/maximum checks.

shehzan10 commented:
> Check.typeOf.number(row, 'row').lessThan(3);
> Check.typeOf.number(row, 'row').lessThanOrEqual(3);
> Check.typeOf.number(row, 'row').lessThan(3).greaterThan(2);

I quite like this. I think its very verbose (LT, LTE etc would have worked too), but that's the Cesium coding style, so I think that's the way to go.

erikmaarten commented:
@shehzan10 agreed, it's a little bit verbose but it's the option I prefer so far too.

mramato commented:
Verbosity leads to clarity.

These functions should have the same names we already use in other places (for example JulianDate), lessThan/lessThanOrEquals/greaterThan/greaterThanOrEquals

I'm not big on chaining, but I'm okay with it in this case.  One minor concern is function call overhead, but we can always optimize if any hotspots show up in the code.

Let's not get lost on the real goals here: 

* Less code needed for error checking.
* Easier to read code for developers.
* Clear and consistent error messages for end users. 

If a manual if check is a better solution to the above than a helper function, we should use it when it makes sense, but I assume that case will be rare.

I agree that getting rid of maximum/minimum in favor of the above lessThan/greaterThan is the right call here.

erikmaarten commented:
After thinking a bit more about this and how it would be implemented, I think this would require using closures or bound functions (or function creators) to capture the original variable tested and its name. Binding will be a lot slower than what we currently have, though it may be worth trying if we think this is important enough. I'm not sure about the impact of using closures, but it's also likely to be slower.

Was this what you meant when you mentioned function call overhead @mramato? When I profiled the Check, a few extra Check calls didn't make a big difference, but a lot of function binding and closures may be worse. Do you think it's worth trying this to see if it works out in practice? If so I can write the code and do some light profiling to see if it seems reasonable.

mramato commented:
These checks should be as simple as possible, so I would avoid extra closures or binding, I'm not sure I understand why they would be needed.  Can you provide an example of the exact problem you are trying to solve?

On some JavaScript implementations, simply calling empty functions thousands of times a frame creates enough overhead to slow down Cesium.  So I would prefer that we implement this in a way so that checking each input variable is a single function call (or at most two, if we really want separate ones for ranges).

lilleyse commented:
My vote is to not use chaining or range functions, but just use `lessThan/lessThanOrEquals/greaterThan/greaterThanOrEquals`

erikmaarten commented:
@mramato With the interface suggested above, `Check.typeOf.number(row, 'row').lessThan(3);`, I'd need to somehow capture the arguments to the first function in order to not have to retype them in the second function call, `lessThan(3)`.

Using bind it could look this:
```js
Check.typeOf.number = function (test, name) {
    // ...
    return {
        lessThan: lessThan.bind(null, test, name)
    };
};
```
And with a nested function with the variables of the first function in its closure, something like this:
```js
Check.typeOf.number = function (test, name) {
    // ...
    return {
        lessThan: function (limit) {
            if (test >= limit) {
                throw new DeveloperError(name + ' expected to be less than ' + limit + ', was ' + test);
            }
        }
    };
};
```
So unless there's another way that I'm missing here we probably can't go down that road.

> My vote is to not use chaining or range functions, but just use `lessThan/lessThanOrEquals/greaterThan/greaterThanOrEquals`

@lilleyse sounds good. @mramato do you agree with this?

mramato commented:
> @mramato With the interface suggested above, Check.typeOf.number(row, 'row').lessThan(3);, I'd need to somehow capture the arguments to the first function in order to not have to retype them in the second function call, lessThan(3).

My preference is that instead of relying on chained calls, we just introduce
```javascript
Check.typeOf.number.lessThan('row', row, 3);
```

And `lessThan` is the one that checks if `row` is a number.  Ideally, each parameter check would only involve a single function call (in the calling code), which will make it easier to read and maintain.  It's also much more efficient than closures or bind, which are notoriously slow on some platforms.

erikmaarten commented:
If we do
```js
Check.typeOf.number.lessThan('row', row, 3);
Check.typeOf.number.greaterThan('row', row, 1);
```
and put the typeOf number check inside `lessThan` and `greaterThan` we'd end up doing the same typeOf number check twice in these cases though.

lilleyse commented:
The first way seems ok to me despite the duplicate typeof check.

erikmaarten commented:
OK. So if everyone's fine with that I think we'll go with the following:

```js
Check.typeOf.number.lessThan('row', row, 3);
Check.typeOf.number.greaterThan('row', row, 1);
```
where
```js
Check.typeOf.number.lessThan = function (name, test, limit) {
    Check.typeOf.number(test, name);
    // now check lessThan condition
}
```

One more thing -- the parameter order here is `variableName, value, limitArgument` whereas it's `value, variableName` in the other Check methods. I think the former makes more sense, especially when having a `limitArgument`, but should we change this for consistency?

lilleyse commented:
I don't really have a strong opinion, but now is definitely the best time to make major updates if you prefer the name going first.

mramato commented:
I agree with @lilleyse consistency is what matters.

erikmaarten commented:
@lilleyse @mramato I'll go ahead and update this then so it makes sense and is consistent.

erikmaarten commented:
@lilleyse @mramato @shehzan10 I wrote a few regex matchers and had the parameter order switched so that the name comes first, and then I added `lessThan`, `lessThanOrEquals`, `greaterThan` and `greaterThanOrEquals`. I also removed some redundant typeof number checks when switching the old `Check.numeric` to the corresponding `Check.typeOf.number.` `greater...` / `less...` while I was at it and already had a bunch of matcher patterns.

Please review.

lilleyse commented:
I made some small style tweaks, also addressed issues in https://github.com/AnalyticalGraphicsInc/cesium/pull/4881#discussion_r96715080 and https://github.com/AnalyticalGraphicsInc/cesium/pull/4835#discussion_r95262030

@austinEng / @shehzan10 were there other cases that were left alone because `greaterThan` and `leftThan` didn't exist yet?

lilleyse commented:
I like the new approach - it all looks good to me.

shehzan10 commented:
@erikmaarten I think the new changes look good. And I think you picked out all the left over `DeveloperError`s from the files that have already been changed. 👍

lilleyse commented:
Would anyone else like to review this? It looks good to me.

mramato commented:
Looking at it now.

mramato commented:
Thanks @erikmaarten this is a big improvement.
